### DISCUSSION TITLE
Use of grange() vs reduce() functions to obtain common peaks in integration of 10XMultiome Datasets

### DISCUSSION THREAD

**Discussion by sylestiel** [stuart-lab/signac#1345] [Q&A] [answered]
Hi @timoast 

I would like some clarification with the workflow on the integrated analysis of 10X Multiome datasets.

You have shown an instance (https://stuartlab.org/signac/articles/integrate_atac.html) where you are using a multiome dataset  to quantify peaks in the scATAC-seq dataset. However, when it comes to  integration of 2 multiome datasets when should we generate the unified peak set (so called combined peaks). There is no instance where we use the FeatureMatrix() function in the Multiome Vignette.  
Can you help clarify **when and where** to use the grange(object)  vs. reduce(object) function with respect to a
nalyzing the same features in the integration of Multiome datasets. The grange function was used as follows: <img width="546" alt="Screen Shot 2023-02-24 at 12 33 17 PM" src="https://user-images.githubusercontent.com/64283689/221262210-776e43c5-bb0a-49c8-87af-fb526a6afbae.png">
 Also, if you could advice as to when to generate the common peak set. Does it precede the CreateChromatinAssay.

Thanks!

**Comment by timoast** [comment, marked as answer]
The `granges()` function will extract the genomic ranges from an existing object.
The `reduce()` function will take a set of genomic ranges and combine them so that there are no overlapping ranges.

When combining multiple single-cell chromatin datasets, it's important that you're using the same set of genomic ranges across all the different datasets, otherwise there's no way to really compare them (for example, running differential accessibility). How you want to define that set of common genomic ranges is up to you, and we provide examples of a few different approaches in the vignettes: choosing the genomic ranges for one dataset and quantifying those ranges for all other datasets, using the `reduce` function to create a merged set of ranges from several objects, or performing peak calling using data from all objects.

Whichever approach you choose to define a set of genomic regions, this needs to be done before quantifying the regions, and before creating the chromatin assay.

**Comment by sylestiel** [comment]
@timoast 

Making sure that I'm doing this correctly. Can you confirm if this approach is correct?

1) If I read in 2 multiome datasets that were processed  **separately**. Do I first call in the .rds files. Merge the two datasets. Process the merged file. ### Then proceed to integrate them using the combined peaks generated using the reduce() function?

2) As an aside I used the peaks.bed file from the Cell Ranger output to get the combined peaks. 
Although in processing of the individual datasets I used CallPeaks() using macs2. 
Would I need to run
Would I need to run
peaks <- [keepStandardChromosomes](https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html)(peaks, pruning.mode = "coarse")
peaks <- [subsetByOverlaps](https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html)(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)
on the combined peaks?

<img width="939" alt="Screen Shot 2023-03-02 at 5 01 55 PM" src="https://user-images.githubusercontent.com/64283689/222579631-e27ebb10-ca3d-46a6-b52f-369305766d78.png">



Thanks!

**Reply by timoast** [reply]
Defining the common set of genomic regions needs to be done before merging objects

**Comment by sylestiel** [comment]
Hi@timoast,

 I did that. 

This is how I did it.

```{r}
# read in peak sets
peaks.s1 <- read.table(
  file = "./ s2/outs /atac_peaks.bed",
  col.names = c("chr", "start", "end")
)
peaks.s2 <- read.table(
  file = "./ s2/outs/atac_peaks.bed",
  col.names = c("chr", "start", "end")
)
```

```{r}
# convert to genomic ranges
gr.s1 <- makeGRangesFromDataFrame(peaks.s1)
gr.s2 <- makeGRangesFromDataFrame(peaks.s2)

```

```{r}
# Create a unified set of peaks to quantify in each dataset
combined.peaks <- reduce(x = c(gr.s1, gr.s2))

```

```{r}
# Filter out bad peaks based on length
peakwidths <- width(combined.peaks)
combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]
combined.peaks
```

```{r}
blacklist_mm10
```

```{r}
# remove peaks on nonstandard chromosomes and in genomic blacklist regions
combined.peaks <- keepStandardChromosomes(combined.peaks, pruning.mode = "coarse")
combined.peaks <- subsetByOverlaps(x = combined.peaks, ranges = blacklist_mm10, invert = TRUE)
```
**I'm getting an error at the end that goes as follows:
Error in idx[i, ] <- res[[i]][[1]] : 
  number of items to replace is not a multiple of replacement length
In addition: Warning message:
In RunIntegration(filtered.anchors = filtered.anchors, normalization.method = normalization.method,  :
  Number of anchors is less than k.weight. Lowering k.weight for sample pair.**

How can I address this issue?  

Thanks!
